x/evm/keeper/state_transition.go

```diff
@@ -349,8 +349,13 @@ func (k *Keeper) ApplyMessageWithConfig(ctx sdk.Context, msg core.Message, trace
 		return nil, sdkerrors.Wrap(types.ErrCallDisabled, "failed to call contract")
 	}
 
+	if !commit {
+		ctx, _ = ctx.CacheContext()
+	}
```

Comment:
Why this code added?

Reply:
precompiled contract also run when estimate gas. estimating gas is not actual execution and state should not be committed.

Reply:
@AgwaB 

Q) ok. but if we create NewEVM with cached context and committed at Line#418, that change is not reflected on the main state?

Reply:
Good point. After hearing you, this could be a vulnerability. If the context input as input is cache and commit is true, there seems to be a problem. Could you please check the additional commits?

Reply:
@AgwaB 
I don't understand the meaning of the code:
```go
if _, ok := ctx.MultiStore().(store.CacheMultiStore); ok && commit; {
    return error
}
```
Could please explain it?

Reply:
As is in the comments. We don't know input parameter `ctx sdk.Context` is cache or not, so should be checked. When a commit is requested, it shouldn't be cached because it needs to be saved to storage.